geopandas/tools/overlay.py

```diff
@@ -76,15 +69,19 @@ def _overlay_difference(df1, df2):
     Overlay Difference operation used in overlay function
     """
     # Spatial Index to create intersections
-    spatial_index = df2.sindex
-    bbox = df1.geometry.apply(lambda x: x.bounds)
-    sidx = bbox.apply(lambda x: list(spatial_index.intersection(x)))
+    df1_idx, df2_idx = df2.sindex.query_bulk(
+        df1.geometry, predicate="intersects", sort=True
+    )
+    df2_geoms = (
+        df2.geometry.values[df2_idx[df1_idx == idx]]
+        if idx in np.unique(df1_idx)
+        else []
+        for idx in range(df1.geometry.size)
+    )
```

Comment:
Can anyone think of a better way to do this using numpy? It's just juggling indexing. Even better if we could avoid looping over the entire df1 altogether and just loop over the indexes in `np.unique(df1_idx)`.